Refactoring and many other changes

msp1974 opened the pull request:
@maciej-or I have taken a look at how this integration is structured and some of the issues raised and done some retructuring which I think (hope!) will both resolve a number of the issues but also make simpler to manage/maintain/enhance going forward.

As this is your great work, please don't think I am trying to take over but want to contribute to a great integration that I use and not create other forks.

I have created a document in this PR called CHANGES.md to explain what I have done and also been running this in the following scenarios without issue:

1. DVR with 4 analog cameras
2. DVR with 4 analog cameras and 1 IP camera
3. DVR with 4 analog cameras as 1 instance and a single IP camera as another instance.

I hope you can acceprt this PR and the code changes are understandable but please reach out if there are any bits to discuss.  In the CHANGES.md document, there are some TODOs which I will also work on before merging this but wanted to see if you were happy with this PR first and then it can be finalised.

Mark

msp1974 commented:
OK @maciej-or , this should now be ready to test.  The version number will need updating before release if you are happy with it.  I have also updated the changes document to reflect that the TODOs/Known issues are now resolved and total changes made.

maciej-or commented on custom_components/hikvision_next/isapi.py:
Do you have ISAPI documentation? If not I could share. Maybe there we will find better way?

maciej-or commented on custom_components/hikvision_next/notifications.py:
```diff
@@ -52,60 +60,127 @@ async def post(self, request: web.Request):
+            raise ValueError(f"Unexpected event Content-Type {content_type_header}")
+        return xml
+
+    def parse_event_notification(self, xml: str) -> AlertInfo:
```
because parse_event_notification contains keys related to ISAPI  better 
keep it in isapi.py

maciej-or reviewed (commented):
Thanks for great contribution. I really like the refactoring and diagnostic data. I tested integration on my setup. NVR works perfectly, standalone ip camera fails with one request. I will investigate that soon and let you know. I'm going to install this branch on my main HA and test for a couple of days.
The PR generally looks good to me, but please follow HA style guidelines https://developers.home-assistant.io/docs/development_guidelines
I have installed development environment according to https://developers.home-assistant.io/docs/development_environment and IDE shows all rules violations.

msp1974 commented:
OK, I have fixed linting errors (I was using a non dev-container setup which doesn't show these) and the other items listed.  I have also removed the HA event firing for now as need to think more through how to do with multiple instances of the integration configured.  Will likely work on this for another PR if thats ok.

Did you find the error with the single IP camera?

maciej-or commented:
Cool, I will manage my single camera issue and we could release. The PR becomes massive so yea create a new branch for next updates.
